libcxx/cmake/config-ix.cmake

```diff
@@ -38,9 +38,13 @@ check_cxx_compiler_flag(-nolibc CXX_SUPPORTS_NOLIBC_FLAG)
 # required during compilation (which has the -nostdlib++ or -nodefaultlibs). libc is
 # required for the link to go through. We remove sanitizers from the
 # configuration checks to avoid spurious link errors.
+#
+# Adding flags to CMAKE_REQUIRED_FLAGS will include the flags both when testing
+# compilation of C and C++. Therefore test to make sure that the flags are
+# supported by the C compiler driver, before deciding to include them.
 
-check_cxx_compiler_flag(-nostdlib++ CXX_SUPPORTS_NOSTDLIBXX_FLAG)
-if (CXX_SUPPORTS_NOSTDLIBXX_FLAG)
+check_c_compiler_flag(-nostdlib++ C_SUPPORTS_NOSTDLIBXX_FLAG)
```

Comment:
Do we compile any C sources in libc++? Why do we care what flags the C compiler supports for libc++ and libc++abi? (I understand why it matters for libunwind where we do have C sources).

The reason I'm asking is that there are some simplifications we can make if we assume that all supported compilers accept `-nostdlib++`, and this patch makes it impossible to make that assumption.

Reply:
The problem is that even if we don’t use any C sources outright, cmake uses C compilation tests when you do a check for whether a library exists. And these are subtly broken if building with GCC right now. (To make matters worse, not all GCC versions are affected. But that’s a bug - GCC intends to reject these options in the C frontend.) IIRC one of the issues mentioned in the linked bug was about library detection in libcxxabi.

Reply:
> The problem is that even if we don’t use any C sources outright, cmake uses C compilation tests when you do a check for whether a library exists.

Do you mean these checks? https://github.com/llvm/llvm-project/blob/aa02b76b1a4c2df1a701fbd8a8378d1cd946e70a/libcxx/cmake/config-ix.cmake#L130

Reply:
Yes, exactly those; any of those basic compiler tests, `check_library_exists` (used a couple of times in libcxxabi and libcxx), `check_include_file` (not used in libcxxabi/libcxx, but used in openmp, and also struck by the same issue, in #90332), are executed as plain C compilation, and they will fail if we have stuck C++-only flags in `CMAKE_REQUIRED_FLAGS`.

Reply:
Ping